sdk/core/azure-core/inc/azure/core/http/websockets/win_http_websockets_transport.hpp

```diff
@@ -0,0 +1,139 @@
+
+#include "azure/core/context.hpp"
+#include "azure/core/http/http.hpp"
+#include "azure/core/http/transport.hpp"
+#include "azure/core/http/websockets/websockets_transport.hpp"
+#include "azure/core/http/win_http_transport.hpp"
+#include <memory>
+#include <mutex>
+
+namespace Azure { namespace Core { namespace Http { namespace WebSockets {
+
+  /**
+   * @brief Concrete implementation of a WebSocket Transport that uses WinHTTP.
+   */
+  class WinHttpWebSocketTransport : public WebSocketTransport, public WinHttpTransport {
+
+    Azure::Core::Http::_detail::unique_HINTERNET m_socketHandle;
+    std::mutex m_sendMutex;
+    std::mutex m_receiveMutex;
+
+    // Called by the
+    void OnUpgradedConnection(
+        Azure::Core::Http::_detail::unique_HINTERNET const& requestHandle) override;
+
+  public:
+    /**
+     * @brief Construct a new WinHTTP WebSocket Transport.
+     *
+     * @param options Optional parameter to override the default options.
+     */
+    WinHttpWebSocketTransport(WinHttpTransportOptions const& options = WinHttpTransportOptions())
+        : WinHttpTransport(options)
+    {
+    }
+    /**
+     * @brief Implements interface to send an HTTP Request and produce an HTTP RawResponse
+     *
+     * @param request an HTTP Request to be send.
+     * @param context A context to control the request lifetime.
+     *
+     * @return unique ptr to an HTTP RawResponse.
+     */
+    virtual std::unique_ptr<RawResponse> Send(Request& request, Context const& context) override;
+
+    /**
+     * @brief Indicates if the transports natively websockets or not.
+     *
+     * @details For the WinHTTP websocket transport, the WinHTTP API supports websockets.
+     */
+    virtual bool HasBuiltInWebSocketSupport() override { return true; }
+
+    /**
+     * @brief Close the underlying WebSocket handle.
+     *
+     */
+    virtual void Close() override;
+
+    // Native WebSocket support methods.
+    /**
+     * @brief Gracefully closes the WebSocket, notifying the remote node of the close reason.
+     *
+     * @details Not implemented for CURL websockets because CURL does not support native websockets.
+     *
+     * @param status Status value to be sent to the remote node. Application defined.
+     * @param disconnectReason UTF-8 encoded reason for the disconnection. Optional.
+     * @param context Context for the operation.
+     *
+     */
+    virtual void NativeCloseSocket(uint16_t, std::string const&, Azure::Core::Context const&)
+        override;
+
+    /**
+     * @brief Retrieve the information associated with a WebSocket close response.
+     *
+     * Should only be called when a Receive operation returns WebSocketFrameType::CloseFrameType
+     *
+     * @param context Context for the operation.
+     *
+     * @returns a tuple containing the status code and string.
+     */
+    virtual NativeWebSocketCloseInformation NativeGetCloseSocketInformation(
+        Azure::Core::Context const& context) override;
```

Comment:
why some of these oveerides have params, and some don't , the rest looks the same

Reply:
`override` tells the C++ compiler that this function is a method override of a base class method. Some of the method overrides in the file have method bodies, others do not.